src/lib.rs

```diff
@@ -58,18 +58,42 @@ use std::eprintln as log_error;
 #[derive(Debug)]
 pub struct WaylandSource<D> {
     queue: EventQueue<D>,
-    fd: Generic<RawFd>,
+    fd: Generic<FdWrapper<RawFd>>,
     read_guard: Option<ReadEventsGuard>,
+    /// Calloop's before_will_sleep method allows
+    /// skipping the sleeping by returning a `Token`.
+    /// We cannot produce this on the fly, so store it here instead
+    fake_token: Option<Token>,
+    // Some calloop event handlers don't support error handling, so we have to store the error
+    // for a short time until we reach a method which allows it
+    stored_error: Result<(), io::Error>,
 }
 
 impl<D> WaylandSource<D> {
     /// Wrap an [`EventQueue`] as a [`WaylandSource`].
-    pub fn new(queue: EventQueue<D>) -> Result<WaylandSource<D>, WaylandError> {
+    ///
+    /// If this returns None, that means that acquiring a read guard failed.
+    /// See [EventQueue::prepare_read] for details
+    /// This guard is only used to get the wayland file descriptor
+    #[must_use]
+    pub fn new(queue: EventQueue<D>) -> Option<WaylandSource<D>> {
         let guard = queue.prepare_read()?;
-        let fd = Generic::new(guard.connection_fd().as_raw_fd(), Interest::READ, Mode::Level);
+        let fd = Generic::new(
```

Comment:
Hm, what's the reason we try to acquire the guard when creating the source? I know that it was like that before, but what is the reason, @elinorbgr ? We won't do any polling here, so it's a bit weird, just to check that the FD is alive?

Reply:
In this repository, it's entirely because `wayland-backend` chooses not to expose the fd through any mechanism other than the guard, as far as I can tell. The `fd` should still be valid any time the event queue (and hence the `Connection`) is around, although we're not given that guarantee.

Reply:
Indeed, though https://github.com/Smithay/wayland-rs/pull/656 is adding the necessary APIs

Reply:
https://github.com/Smithay/wayland-rs/pull/656 lets us do this in a clean and safe way with simply `let queue = Generic::new(queue, Interest::READ, Mode::Level);`. Then using `self.queue.file` to access the contained `EventQueue`.

https://github.com/Smithay/client-toolkit/pull/403/files#diff-8d59229f4c6aea492f0cbceaafc305fcade8dbe513b86e6466feed5a0b2b3107 has a version doing that.

Reply:
We'd need to wait for a patch release, but I'd rather do so.